agent/grpc/stats_test.go

```diff
@@ -64,23 +65,42 @@ func TestHandler_EmitsStats(t *testing.T) {
 	// Wait for the server to stop so that active_streams is predictable.
 	require.NoError(t, g.Wait())
 
+	// Occasionally the active_stream=0 metric may be emitted before the
+	// active_conns=0 metric. The order of those metrics is not really important
+	// so we sort the calls to match the expected.
```

Comment:
This handling fixes one flake, but there is still one left:
```
--- FAIL: TestHandler_EmitsStats (0.00s)
    stats_test.go:88: assertion failed: values are not equal
        --- expected
        +++ actual
          []grpc.metricCall{
                {key: {"testing", "grpc", "server", "active_conns"}, val: 1},
        -       {key: []string{"testing", "grpc", "server", "active_streams"}, val: 1},
                {key: {"testing", "grpc", "server", "active_conns"}},
        +       {key: []string{"testing", "grpc", "server", "active_conns"}, val: 1},
        +       {key: []string{"testing", "grpc", "server", "active_conns"}, val: 2},
        +       {key: []string{"testing", "grpc", "server", "active_conns"}, val: 1},
        +       {key: []string{"testing", "grpc", "server", "active_conns"}},
        +       {key: []string{"testing", "grpc", "server", "active_streams"}, val: 1},
                {key: {"testing", "grpc", "server", "active_streams"}},
          }
```

I haven't been able to locate the problem. If I run this test by itself (`go test -count=1000 -run TestHandler_EmitsStats`) it never flakes, but if I run all the tests in the package it will occasionally fail because there are extra connections made.

From what I can tell (based on this flake, and the comment above in the fakeRPCListener) these servers are outliving the test, and a client from another test can connect to them. I'm not sure how that is possible because we use a unique scheme for each test, and we wait for the server to shutdown in `t.Cleanup`.

Maybe grpc is doing something I don't understand in background goroutines? Or maybe `t.Cleanup` is still running when the next test is started?

It takes a few hundred runs to reproduce the flake, so hopefully we don't see them very often in CI.

Reply:
Good detective work! Sounds like this is a good point to leave it for now and we can invest more if we see it causing CI issues.